session/sql_migration_test.go

```diff
@@ -0,0 +1,738 @@
+
+				label = fmt.Sprintf("linkedSession%d", j+1)
+
+				activeSess, err = kvStore.NewSession(
+					ctx, label, activeSess.Type,
+					time.Unix(1000, 0),
+					activeSess.ServerAddr, opts...,
+				)
+				require.NoError(t, err)
+			}
+		}
+
+		// Finally, we shift the active session to a random state.
+		// As the state we set may be a state that's no longer set
+		// through the current code base, or be an illegal state
+		// transition, we use an alternative test state shifting method
+		// that doesn't check that we transition the state in the legal
+		// order.
+		err = kvStore.shiftStateUnsafe(ctx, activeSess.ID, lastState(i))
+		require.NoError(t, err)
+	}
+}
+
+// macaroonType returns a macaroon type based on the given index by taking the
+// index modulo 6. This ensures an approximately equal distribution of macaroon
+// types.
+func macaroonType(i int) Type {
+	switch i % 6 {
+	case 0:
+		return TypeMacaroonReadonly
+	case 1:
+		return TypeMacaroonAdmin
+	case 2:
+		return TypeMacaroonCustom
+	case 3:
+		return TypeUIPassword
+	case 4:
+		return TypeAutopilot
+	default:
+		return TypeMacaroonAccount
+	}
+}
+
+// lastState returns a state based on the given index by taking the index modulo
+// 5. This ensures an approximately equal distribution of states.
+func lastState(i int) State {
+	switch i % 5 {
+	case 0:
+		return StateCreated
+	case 1:
+		return StateInUse
+	case 2:
+		return StateRevoked
+	case 3:
+		return StateExpired
+	default:
+		return StateReserved
+	}
+}
+
+// randomPrivacyFlags returns a random set of privacy flags.
+func randomPrivacyFlags() PrivacyFlags {
+	allFlags := []PrivacyFlag{
+		ClearPubkeys,
+		ClearChanIDs,
+		ClearTimeStamps,
+		ClearChanInitiator,
+		ClearHTLCs,
+		ClearClosingTxIds,
+		ClearNetworkAddresses,
+	}
+
+	var privFlags []PrivacyFlag
+	for _, flag := range allFlags {
+		if rand.Intn(2) == 0 {
+			privFlags = append(privFlags, flag)
+		}
+	}
+
+	return privFlags
+}
+
+// randomFeatureConfig returns a random feature config with a random number of
+// features. The feature names are generated as "feature0", "feature1", etc.
+func randomFeatureConfig() FeaturesConfig {
+	featureConfig := make(FeaturesConfig)
+	for i := 0; i < rand.Intn(10); i++ {
+		featureName := fmt.Sprintf("feature%d", i)
+		featureValue := []byte{byte(rand.Int31())}
+		featureConfig[featureName] = featureValue
+	}
+
+	return featureConfig
+}
+
+// randomMacaroonRecipe returns a random macaroon recipe with a random number of
+// caveats and permissions. The returned macaroon recipe may have nil set for
+// either the caveats or permissions, but not both.
+func randomMacaroonRecipe() Option {
+	var (
+		macCaveats []macaroon.Caveat
+		macPerms   []bakery.Op
+	)
+
+	loopLen := rand.Intn(10) + 1
+
+	if rand.Intn(2) == 0 {
+		for i := 0; i < loopLen; i++ {
+			var macCaveat macaroon.Caveat
+
+			// We always have a caveat.Id, but the rest are
+			// randomized if they exist or not.
+			macCaveat.Id = randomBytes(rand.Intn(10) + 1)
```

Comment:
Let me know if this **is not true** in for real world data, as migrations WILL fail if we don't set a `caveat.Id` when the `caveats` param of `WithMacaroonRecipe` is not `nil`.

Reply:
Could potentially override this as well in the new `overrideMacaroonRecipe` as well. But I do not think this scenario can happen for real world data. But open to implementing that as well if you think it's worth doing it just to be sure.